src/map/achievement.c

```diff
@@ -0,0 +1,1693 @@
+
+	/* Status Types */
+	equal = (objective->unique.status_type && objective->unique.status_type == criteria->unique.status_type);
+
+	/* Item Type */
+	equal = (objective->item_type && (objective->item_type&criteria->item_type) != 0);
+
+	/* Job Ids */
+	for (i = 0; i < VECTOR_LENGTH(objective->jobid); i++) {
+		ARR_FIND(0, VECTOR_LENGTH(criteria->jobid), j, VECTOR_INDEX(criteria->jobid, j) == VECTOR_INDEX(objective->jobid, i));
+		equal = (j < VECTOR_LENGTH(criteria->jobid));
+	}
+
+	return equal;
+}
+
+/**
+ * Validates an Achievement Objective of similar types.
+ * @param sd          as a pointer to the map session data.
+ * @param type        as the type of the achievement.
+ * @param criteria    as the criteria of the objective (mob id, job id etc.. 0 for no criteria).
+ * @param progress    as the current progress of the objective.
+ * @return total number of updated achievements on success, 0 on failure.
+ */
+int achievement_validate_type(struct map_session_data *sd, const enum achievement_types type, const struct achievement_objective *criteria, const int progress, bool additive)
+{
+	int i = 0, total = 0;
+	struct achievement *ach = NULL;
+
+	nullpo_ret(sd);
+	Assert_ret(progress > 0);
+
+	if (type == ACH_QUEST) {
+		ShowDebug("achievement_validate_type: ACH_QUEST is not handled by this function. (use achievement_validate())\n");
+		return 0;
+	} else if (type >= ACH_TYPE_MAX) {
+		ShowDebug("achievement_validate_type: Invalid Achievement Type %d! (min: %d, max: %d)\n", (int)type, (int)ACH_QUEST, (int)ACH_TYPE_MAX - 1);
+		return 0;
+	}
+
+	/* Loop through all achievements of the type, checking for possible matches. */
+	for (i = 0; i < VECTOR_LENGTH(achievement->category[type]); i++) {
+		int j = 0;
+		bool updated = false;
+		struct achievement_data *ad = NULL;
+
+		if ((ad = achievement->get(VECTOR_INDEX(achievement->category[type], i))) == NULL)
+			continue;
+		
+		for (j = 0; j < VECTOR_LENGTH(ad->objective); j++) {
+			// Check if objective criteria matches.
+			if (achievement->check_criteria(&VECTOR_INDEX(ad->objective, j), criteria) == false)
+				continue;
+			// Criteria passed, check if not completed and update progress.
+			if ((ach = achievement->find_or_create(sd, ad->id, false)) == NULL
+				|| (!ach->completed_at && ach->objective[j] < VECTOR_INDEX(ad->objective, j).goal)) {
+				if (additive)
+					achievement->progress_add(sd, ad, j, progress);
+				else
+					achievement->progress_set(sd, ad, j, progress);
+				updated = true;
+			}
+		}
+
+		if (updated) total++;
+	}
+
+	return total;
+}
+
+/**
+ * Validates any achievement's specific objective index.
+ * @param sd       pointer to the session data.
+ * @param aid      ID of the achievement.
+ * @param index    index of the objective.
+ * @param progress progress to be added towards the goal.
+ */
+bool achievement_validate(struct map_session_data *sd, const int aid, const unsigned int index, const int progress, bool additive)
```

Comment:
> Well, correct but there are no pass-by references to this function. So if you're passing a const, it would prevent any future statements from changing the value of the variable because we absolutely don't need it to. (Including plugin hooks)

In general (at least in Hercules) the only meaning of the `const` keyword is as a guarantee to the caller. And the caller doesn't care about what the called function does with the data (i.e. if it wishes to, it can make a copy of it, or three, and change them to whatever other value). A function signature mustn't limit, or even care aboout, what a function does with its own private copy of the data.

> Even in the case of plugins, I can't see why the achievement and objective Id would need to be changed in a pre-hook.

A plugin author may see why, and we aren't here to limit their freedom of expression.

> Since changing the first two arguments in this function makes as much sense as wanting to change the pointer to a session data in a function that uses it as a key argument (example: pc_delitem - where in its pre/post-hook we allow the changing of the pointer since it's a non-const).

Which is a perfectly valid thing for a plugin to do. If a plugin wants to remap IDs differently for any reason, it should be able to. Just like a plugin can prevent this function from being called at all by stopping the hook before the function is called (as long as it respect its contract with the caller). If a plugin wants to pass a different `sd` to `pc_delitem` why would you want to stop it? The plugin might be mapping a player's inventory to another player's for whatever reason the author wishes, and might still want to call `pc_delitem` with the swapped pointer -- and this won't affect the caller in any way, since the change only happens in the pre-hook and inside the called function.

> If your design idea for HPM incorporates the changing of anything, anywhere, then I'll remove the consts. I just wanted an explanation that made sense because it is not wrong to pass scalars as const unless they are allowed to be changed within the scope of the function.

Not "anything, anywhere". Using const is still a very good thing to do for pointers (and with this I mean pointers to const data, not const pointers to mutable data -- those would be as useless as const scalars). Const only has a reason to exist (in Hercules at least) when it guarantees to the caller that the called function doesn't change the value of the passed variables. I'll make an example on the only meaningful reason I'd ever want to use `const` for, in a function signature:

```

/*
 * These functions may be defined elsewhere and
 * we don't even want to care about their internal
 * implementation details
 */
int fun(const char *foo);
int notsofun(char *foo);
int notfunatall(char * const foo);

int caller(void)
{
    char buf[256] = "foo";

    /* ... */

    // I created or received 'buf' and I validated it, ensuring it contains a meaningful value of any kind

    fun(buf);
    // I still know that the content of buf hasn't changed, and I don't need to re-ensure anything

    notsofun(buf);
    // Now I have no guarantees about anything, because the function I passed this to
    // takes a pointer to mutable data and might have changed *my* data (the caller's data)

    notfunatall(buf);
    // Once again, no guarantee. The called function is telling us that in their implementation
    // they won't change the address their local copy of the pointer points to, but really, who
    // cares? They don't guarantee, though, that they won't mess with the pointed data (which
    // was my data to begin with, not theirs)
}
```

But with scalars this all becomes pointless:

```
/*
 * These functions may be defined elsewhere and
 * we don't even want to care about their internal
 * implementation details
 */
int constfun(const int foo);
int notconstfun(int foo);
int anotherfun(int foo);

/* over-simplified HPMHooking example */
int anotherfun_pre(int *foo)
{
    if (*foo == 1)
        *foo = 42; // And so what? I want the called function to process 42 instead, can't I?
}
int anotherfun_post(int ret, int *foo)
{
    if (*foo == 42 && ret == 1)
        return 0; // Why not
}
int anotherfun_with_hpmhooking(int foo)
{
    int ret_ = anotherfun_pre(&foo);
    /* ... */
    ret_ = anotherfun(foo);
    /* ... */
    return anotherfun_post(ret_, &foo);
}

int caller(void)
{
    int foo = 1;
    const int bar = 1;

    /* ... */
    // I validated or created my data and as such I know the value of both foo and bar.

    constfun(foo);
    constfun(bar);
    notconstfun(foo);
    notconstfun(bar);

    // No difference at all for me to call either function with either const or not const data.
    // Regardless of what I call (by value), I don't care about what they did with the data. But even
    // if their local copy of my data was not const, I know they weren't able to change my copy
    // and that's all I care about.


    anotherfun_with_hpmhooking(foo);
    anotherfun_with_hpmhooking(bar);
    // And even if a plugin needs to play around with the data, I still don't care or want to care.
    // It's working on a copy of the data, so why should I limit it? The plugin knows `anotherfun()`'s
    // internals better than I do, and knows what its doing. It's a plugin that aims to modify
    // modifies that function's behavior after all.
}
```